_posts/2023-07-29-oauth-authentication-scala-http4s.md

```diff
@@ -0,0 +1,672 @@
+---
+title: "Oauth Authentication in Scala with Http4s"
+date: 2023-07-27
+header:
+  image: "/images/blog cover.jpg"
+tags: []
+excerpt: "Learn about OAuth authentication using Http4s."
+toc: true
+toc_label: "In this article"
+---
+
+## Introduction
+
+This is a follow up to the [Two Factor Authentication](https://blog.rockthejvm.com/otp-authentication-scala-http4s/) article that wass written earlier, and as such is the third article in the series.
+
+OAuth which stands for Open Authorization is an open standard framework that allows the user to permit a website or application to interact with another without giving up his or her password.
+
+1. When the user tries to log into app1, the user is redirected to an authorization server owned by app2.
+1. The authorization server provides the user with a prompt, asking the user to grant app1 access to app2 with a list of permissions.
+1. Once the prompt is accepted, the user is redirected back to app1 with a single-use authorization code.
+1. app1 will respond to app2 with the same authorization code, a client id, and a client secret.
+1. The authorization server on app2 will respond with a token id and an access token
+1. app1 can now request the user's information from app2's API using the access token.
+
+The OAuth standard is defined under [RFC 6749](https://www.ietf.org/rfc/rfc6749.txt), here we'll find an in-depth explanation of how the framework works.
+
+## Accessing the Github API through OAuth.
+
+### Setting Up.
+
+In this section, we'll build an application that connects to GitHub using OAuth and requests user's information using the GitHub API.
+
+To build this application we will need to add the following to our build.sbt file:
+
+```scala
+val scala3Version = "3.2.2"
+val Http4sVersion = "0.23.18"
+val CirisVersion = "3.2.0"
+val CirceVersion = "0.14.1"
+
+val emberServer =     "org.http4s"                %% "http4s-ember-server" % Http4sVersion
+val emberClient =     "org.http4s"                %% "http4s-ember-client" % Http4sVersion
+val http4sDsl =       "org.http4s"                %% "http4s-dsl"          % Http4sVersion
+val ciris =           "is.cir"                    %% "ciris"               % CirisVersion
+val cirisCirce =      "is.cir"                    %% "ciris-circe"         % CirisVersion
+val circeLibs =  Seq(
+  "io.circe" %% "circe-core",
+  "io.circe" %% "circe-parser"
+).map(_ % CirceVersion)
+
+  lazy val oauth = project
+  .in(file("oauth"))
+  .settings(
+    name := "oauth",
+    version := "0.1.0-SNAPSHOT",
+    scalaVersion := scala3Version,
+    libraryDependencies ++= Seq(
+      emberServer,
+      emberClient,
+      http4sDsl,
+      http4sCirce,
+      ciris,
+      cirisCirce
+    ) ++ circeLibs
+  )
+```
+
+We'll be using Scala 3 for this tutorial however one can still use Scala 2.13 with minimal code changes.
```

Comment:
Minor phrasing change:
"We use Scala 3 in this tutorial - if you need to use Scala 2.13, you can do so with minimal code changes".

Reply:
Updated